### PR TITLE
Bump libp2p to 0.40.0

### PR DESCRIPTION
**Pull request by kpp**
- What does it do?
  
  Updates libp2p to a 0.40.0 version.

### CODE REVIEW

**Comment by dvdplm** [comment]
@kpp Any news on this? Now that 0.40 is out this could use an update?

**Comment by kpp** [comment]
Yes, I will update it.
News: there are burn-ins running - https://grafana.parity-mgmt.parity.io/explore?orgId=1&left=%5B%22now-1h%22%2C%22now%22%2C%22loki-polkadot.stakingops-logs%22%2C%7B%22expr%22%3A%22%7Bhost%3D%5C%22polkadot-burnin-fullnode-1%5C%22%7D%22%7D%5D , we should review them.

**Comment by tomaka** [comment]
I think that some methods have been added to the `NetworkBehaviour` trait in libp2p 0.40
https://docs.rs/libp2p/0.40.0/libp2p/swarm/trait.NetworkBehaviour.html

Could you make sure that all our implementations of this trait have all the methods when necessary?
I see that `inject_address_change` is missing in `discovery.rs`.

**Comment by kpp** [comment]
> `inject_address_change` is missing in discovery.rs

It was introduced in libp2p-swarm in 0.20.0 [2020-07-01].

I made a diff of methods of NetworkBehaviour:

```diff
$ diff 0.39.txt 0.40.txt 
3d2
< inject_addr_reach_failure
13a13
> inject_listen_failure
```

So only NetworkBehaviour::inject_listen_failure should be addressed in this PR. Will fix.

File: client/network/src/discovery.rs
Comment (tomaka) [inline client/network/src/discovery.rs:598]: That's not really future-proof. Maybe in a future version it will do something.
We need a way to deconstruct a `MultiHandler` and libp2p doesn't provide any. When this PR is merged, could you open an issue in libp2p and in Substrate and follow up with a fix?

File: client/network/src/discovery.rs
Comment (tomaka) [inline client/network/src/discovery.rs:693]: Same remark as above

File: client/network/src/discovery.rs
Comment (tomaka) [inline client/network/src/discovery.rs:775]: Could you change this for a `debug_assert!(false, "...")` and ignore the event instead? This way we panic in debug mode but the node will continue to run in release.

File: client/network/src/peer_info.rs
Comment (tomaka) [inline client/network/src/peer_info.rs:198]: Please do it anyway. The fact that Ping and Identify do nothing is an implementation detail.

File: client/network/src/peer_info.rs
Comment (tomaka) [inline client/network/src/peer_info.rs:314]: Same remark.

**Review by tomaka — CHANGES_REQUESTED** [review]

File: client/network/src/discovery.rs
Comment (kpp) [reply client/network/src/discovery.rs:598]: @mxinden can you take a look at this one? ^

**Review by tomaka — APPROVED** [review]
LGTM provided we fix my two remaining review comments in the near future.

File: client/network/src/discovery.rs
Comment (mxinden) [reply client/network/src/discovery.rs:598]: > We need a way to deconstruct a MultiHandler and libp2p doesn't provide any. 

Any road blockers to contributing this upstream (rust-libp2p)?

**Review by bkchr — APPROVED** [review]
Just needs master merge

**Comment by tomaka** [comment]
bot merge

**Pull request merged**
